aptos-move/framework/aptos-stdlib/sources/data_structures/smart_vector.move

```diff
@@ -336,11 +336,124 @@ module aptos_std::smart_vector {
+            i = i + 1
+        }
+    }
+
+    /// Apply the function to a mutable reference to each T in the vector.
+    public inline fun for_each_mut<T>(v: &mut SmartVector<T>, f: |&mut T|) {
+        let i = 0;
+        let len = aptos_std::smart_vector::length(v);
+        while (i < len) {
+            f(aptos_std::smart_vector::borrow_mut(v, i));
+            i = i + 1
+        }
+    }
+
+    /// Apply the function to a reference of each T in the vector with its index.
+    public inline fun enumerate_ref<T>(v: &SmartVector<T>, f: |u64, &T|) {
+        let i = 0;
+        let len = aptos_std::smart_vector::length(v);
+        while (i < len) {
+            f(i, aptos_std::smart_vector::borrow(v, i));
+            i = i + 1;
+        };
+    }
+
+    /// Apply the function to a mutable reference of each T in the vector with its index.
+    public inline fun enumerate_mut<T>(v: &mut SmartVector<T>, f: |u64, &mut T|) {
+        let i = 0;
+        let len = length(v);
+        while (i < len) {
+            f(i, borrow_mut(v, i));
+            i = i + 1;
+        };
+    }
+
+    /// Fold the function over the Ts. For example, `fold(vector[1,2,3], 0, f)` will execute
+    /// `f(f(f(0, 1), 2), 3)`
+    public inline fun fold<Accumulator, T: store>(
+        v: SmartVector<T>,
+        init: Accumulator,
+        f: |Accumulator, T|Accumulator
+    ): Accumulator {
+        let accu = init;
+        aptos_std::smart_vector::for_each(v, |elem| accu = f(accu, elem));
+        accu
+    }
+
+    /// Fold right like fold above but working right to left. For example, `fold(vector[1,2,3], 0, f)` will execute
+    /// `f(1, f(2, f(3, 0)))`
+    public inline fun foldr<Accumulator, T>(
+        v: SmartVector<T>,
+        init: Accumulator,
+        f: |T, Accumulator|Accumulator
+    ): Accumulator {
+        let accu = init;
+        aptos_std::smart_vector::for_each_reverse(v, |elem| accu = f(elem, accu));
+        accu
+    }
+
+    /// Map the function over the references of the Ts of the vector, producing a new vector without modifying the
+    /// original vector.
+    public inline fun map_ref<T1, T2: store>(
+        v: &SmartVector<T1>,
+        f: |&T1|T2
+    ): SmartVector<T2> {
+        let result = aptos_std::smart_vector::new<T2>();
+        aptos_std::smart_vector::for_each_ref(v, |elem| aptos_std::smart_vector::push_back(&mut result, f(elem)));
+        result
+    }
+
+    /// Map the function over the Ts of the vector, producing a new vector.
+    public inline fun map<T1: store, T2: store>(
+        v: SmartVector<T1>,
+        f: |T1|T2
+    ): SmartVector<T2> {
+        let result = aptos_std::smart_vector::new<T2>();
+        aptos_std::smart_vector::for_each(v, |elem| push_back(&mut result, f(elem)));
+        result
+    }
+
+    /// Filter the vector using the boolean function, removing all Ts for which `p(e)` is not true.
+    public inline fun filter<T: store + drop>(
```

Comment:
What about find that accepts a lambda (instead of value)? vector needs one such function too :P

Reply:
vector already has that. I 'll add one.

Reply:
find doesn't work cuz it will trigger the bug  https://github.com/aptos-labs/aptos-core/issues/9717
https://github.com/aptos-labs/aptos-core/issues/6615 again.